src/Authentication/ApiKey.php

```diff
@@ -0,0 +1,379 @@
+<?php
+/**
+ * API key.
+ *
+ * @package SatisPress
+ * @license GPL-2.0-or-later
+ * @since 0.3.0
+ */
+
+declare ( strict_types = 1 );
+
+namespace SatisPress\Authentication;
+
+use function SatisPress\generate_random_string;
+use ArrayAccess;
+use DateTime;
+use DateTimeZone;
+use WP_User;
+use WP_User_Query;
+
+/**
+ * API key class.
+ *
+ * @since 0.3.0
+ */
+final class ApiKey implements ArrayAccess {
```

Comment:
This class does too much, compared to your usual architectural decisions.

There's a mixture of normal and static methods, that:
 - gets meta tokens
 - formats dates
 - creates keys, but then also saves and revokes keys while interacting with an unchangeable storage layer
 - treats itself as both a single item, and as a collection of keys, with find_by_*() methods as well as ArrayAccess.

I think this could be split into at least three classes, if not more.

Reply:
Yeah, I agree it could be split up quite a bit, but was trying to keep it fairly contained and simple for now since it's primarily only used during authentication and I didn't want to over-engineer it. This approach was mostly modeled after the one in the [WP API OAuth 2 plugin](https://github.com/WP-API/OAuth2/tree/master/inc/tokens).

I'm not worried about the storage being swappable since other authentication providers can be implemented. 

The accessor methods are for primary fields or data that may need some sort of handling before being returned. ArrayAccess is just a convenient way for storing extra data that might be useful at some point, but that could be replaced with a couple of generic methods. It's kinda like a JWT that you can add claims to.

The static factory and find methods are more of an active record pattern than anything else, but none of the other objects really need to persist state. If this approach does need to be fleshed out in the future, I figured those particular methods could be deprecated.

I'm not completely opposed to refactoring, but it was nice to get something working somewhat quickly 😄